File: app/src/test/java/com/kickstarter/viewmodels/FacebookConfimationViewModelTest.java
package com.kickstarter.viewmodels;

import android.content.Intent;
import android.support.annotation.NonNull;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.factories.ApiExceptionFactory;
import com.kickstarter.factories.ConfigFactory;
import com.kickstarter.libs.CurrentConfigType;
import com.kickstarter.libs.Environment;
import com.kickstarter.libs.KoalaEvent;
import com.kickstarter.libs.MockCurrentConfig;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.services.MockApiClient;
import com.kickstarter.services.apiresponses.AccessTokenEnvelope;
import com.kickstarter.services.apiresponses.ErrorEnvelope;
import com.kickstarter.ui.IntentKey;

import org.junit.Test;

import java.util.Collections;

import rx.Observable;
import rx.observers.TestSubscriber;

public class FacebookConfimationViewModelTest extends KSRobolectricTestCase {
  private FacebookConfirmationViewModel.ViewModel vm;
  private final TestSubscriber<String> prefillEmail = new TestSubscriber<>();
  private final TestSubscriber<String> signupError = new TestSubscriber<>();
  private final TestSubscriber<Void> signupSuccess = new TestSubscriber<>();
  private final TestSubscriber<Boolean> sendNewslettersIsChecked = new TestSubscriber<>();

  @Test
  public void testPrefillEmail() {
    final ErrorEnvelope.FacebookUser facebookUser = ErrorEnvelope.FacebookUser.builder()
      .id(1).name("Test").email("test@kickstarter.com")
      .build();

    this.vm = new FacebookConfirmationViewModel.ViewModel(environment());
    this.vm.intent(new Intent().putExtra(IntentKey.FACEBOOK_USER, facebookUser));

    this.vm.outputs.prefillEmail().subscribe(this.prefillEmail);

    this.prefillEmail.assertValue("test@kickstarter.com");
    this.koalaTest.assertValues(KoalaEvent.FACEBOOK_CONFIRM, KoalaEvent.USER_SIGNUP);
  }

  @Test
  public void testSignupErrorDisplay() {
    final ApiClientType apiClient = new MockApiClient() {
      @Override
      public @NonNull Observable<AccessTokenEnvelope> registerWithFacebook(final @NonNull String fbAccessToken, final boolean sendNewsletters) {
        return Observable.error(ApiExceptionFactory.apiError(
          ErrorEnvelope.builder().httpCode(404).errorMessages(Collections.singletonList("oh no")).build())
        );
      }
    };

    final Environment environment = environment().toBuilder().apiClient(apiClient).build();
    this.vm = new FacebookConfirmationViewModel.ViewModel(environment);

    this.vm.intent(new Intent().putExtra(IntentKey.FACEBOOK_TOKEN, "token"));
    this.vm.outputs.signupError().subscribe(this.signupError);

    this.koalaTest.assertValues(KoalaEvent.FACEBOOK_CONFIRM, KoalaEvent.USER_SIGNUP);

    this.vm.inputs.sendNewslettersClick(true);
    this.vm.inputs.createNewAccountClick();

    this.signupError.assertValue("oh no");
    this.koalaTest.assertValues(KoalaEvent.FACEBOOK_CONFIRM, KoalaEvent.USER_SIGNUP,
      KoalaEvent.SIGNUP_NEWSLETTER_TOGGLE, KoalaEvent.ERRORED_USER_SIGNUP);
  }

  @Test
  public void testSuccessfulUserCreation() {
    final ApiClientType apiClient = new MockApiClient();

    final Environment environment = environment().toBuilder().apiClient(apiClient).build();
    this.vm = new FacebookConfirmationViewModel.ViewModel(environment);

    this.vm.intent(new Intent().putExtra(IntentKey.FACEBOOK_TOKEN, "token"));
    this.vm.outputs.signupSuccess().subscribe(this.signupSuccess);

    this.vm.inputs.sendNewslettersClick(true);
    this.vm.inputs.createNewAccountClick();

    this.signupSuccess.assertValueCount(1);
    this.koalaTest.assertValues(KoalaEvent.FACEBOOK_CONFIRM, KoalaEvent.USER_SIGNUP,
      KoalaEvent.SIGNUP_NEWSLETTER_TOGGLE, KoalaEvent.LOGIN, KoalaEvent.NEW_USER);
  }

  @Test
  public void testToggleSendNewsLetter_isNotChecked() {
    final CurrentConfigType currentConfig = new MockCurrentConfig();
    currentConfig.config(ConfigFactory.config().toBuilder().countryCode("US").build());
    final Environment environment = environment().toBuilder().currentConfig(currentConfig).build();
    this.vm = new FacebookConfirmationViewModel.ViewModel(environment);

    this.vm.outputs.sendNewslettersIsChecked().subscribe(this.sendNewslettersIsChecked);
    this.sendNewslettersIsChecked.assertValue(false);

    this.vm.inputs.sendNewslettersClick(true);
    this.vm.inputs.sendNewslettersClick(false);

    this.sendNewslettersIsChecked.assertValues(false, true, false);
    this.koalaTest.assertValues(KoalaEvent.FACEBOOK_CONFIRM, KoalaEvent.USER_SIGNUP,
      KoalaEvent.SIGNUP_NEWSLETTER_TOGGLE, KoalaEvent.SIGNUP_NEWSLETTER_TOGGLE);
  }
}

File: app/src/test/java/com/kickstarter/viewmodels/FacebookConfimationViewModelTest.kt
package com.kickstarter.viewmodels

import android.content.Intent
import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.libs.CurrentConfigTypeV2
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.mock.MockCurrentConfigV2
import com.kickstarter.mock.factories.ApiExceptionFactory
import com.kickstarter.mock.factories.ConfigFactory.config
import com.kickstarter.mock.services.MockApiClientV2
import com.kickstarter.services.ApiClientTypeV2
import com.kickstarter.services.apiresponses.AccessTokenEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope
import com.kickstarter.services.apiresponses.ErrorEnvelope.FacebookUser
import com.kickstarter.ui.IntentKey
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subscribers.TestSubscriber
import org.junit.After
import org.junit.Test

class FacebookConfimationViewModelTest : KSRobolectricTestCase() {
    private val prefillEmail = TestSubscriber<String>()
    private val signupError = TestSubscriber<String>()
    private val signupSuccess = TestSubscriber<Unit>()
    private val sendNewslettersIsChecked = TestSubscriber<Boolean>()
    private val disposables = CompositeDisposable()

    @Test
    fun testPrefillEmail() {
        val facebookUser = FacebookUser.builder()
            .id(1L).name("Test").email("test@kickstarter.com")
            .build()

        val vm = FacebookConfirmationViewModel.FacebookConfirmationViewModel(environment())
        vm.provideIntent(Intent().putExtra(IntentKey.FACEBOOK_USER, facebookUser))

        vm.outputs.prefillEmail().subscribe { this.prefillEmail.onNext(it) }.addToDisposable(disposables)

        prefillEmail.assertValue("test@kickstarter.com")
    }

    @Test
    fun testSignupErrorDisplay() {
        val apiClient: ApiClientTypeV2 = object : MockApiClientV2() {
            override fun registerWithFacebook(
                fbAccessToken: String,
                sendNewsletters: Boolean
            ): Observable<AccessTokenEnvelope> {
                return Observable.error(
                    ApiExceptionFactory.apiError(
                        ErrorEnvelope.builder().httpCode(404).errorMessages(listOf("oh no")).build()
                    )
                )
            }
        }

        val environment = environment().toBuilder().apiClientV2(apiClient).build()
        val vm = FacebookConfirmationViewModel.FacebookConfirmationViewModel(environment)

        vm.outputs.signupError().subscribe { this.signupError.onNext(it) }.addToDisposable(disposables)

        vm.provideIntent(Intent().putExtra(IntentKey.FACEBOOK_TOKEN, "token"))
        vm.inputs.sendNewslettersClick(true)
        vm.inputs.createNewAccountClick()

        signupError.assertValue("oh no")
    }

    @Test
    fun testSuccessfulUserCreation() {
        val apiClient: ApiClientTypeV2 = MockApiClientV2()

        val environment = environment().toBuilder().apiClientV2(apiClient).build()
        val vm = FacebookConfirmationViewModel.FacebookConfirmationViewModel(environment)

        vm.outputs.signupSuccess().subscribe { this.signupSuccess.onNext(it) }.addToDisposable(disposables)

        vm.provideIntent(Intent().putExtra(IntentKey.FACEBOOK_TOKEN, "token"))
        vm.inputs.sendNewslettersClick(true)
        vm.inputs.createNewAccountClick()

        signupSuccess.assertValueCount(1)
    }

    @Test
    fun testToggleSendNewsLetter_isNotChecked() {
        val currentConfig: CurrentConfigTypeV2 = MockCurrentConfigV2()
        currentConfig.config(config().toBuilder().countryCode("US").build())
        val environment = environment().toBuilder().currentConfig2(currentConfig).build()
        val vm = FacebookConfirmationViewModel.FacebookConfirmationViewModel(environment)

        vm.outputs.sendNewslettersIsChecked().subscribe { this.sendNewslettersIsChecked.onNext(it) }.addToDisposable(disposables)
        sendNewslettersIsChecked.assertValue(false)

        vm.inputs.sendNewslettersClick(true)
        vm.inputs.sendNewslettersClick(false)

        sendNewslettersIsChecked.assertValues(false, true, false)
    }

    @After
    fun clear() {
        disposables.clear()
    }
}
